### DISCUSSION TITLE
Rework API

### DISCUSSION THREAD

**Discussion by dingo35** [serkri/SmartEVSE-3#163] [Ideas]
Over time the API has grown big and organically; the /settings doc String has expanded to 1500 bytes now, and the end is not in sight.

Now the MQTT client is added, which is another way to interface with the outside world.

In this discussion I would like valued contributors to give their ideas and comments on a new design of the API, resulting in a uniform API acceptable to all and logical for newcomers.

In rethinking this, I tried to sum up some "requirements" for the new API:

- 1 uniform definition, names and hierarchy for RAW, ha-integration, MQTT, endpoints and swagger.xml
- share as much code as possible, e.g. by generating reusable JSON datastructures compile-time; minimal usage of runtime!
- reduce memory usage by working with small chunks (so the current /settings endpoint would be a bad example)
- perhaps in the hierarchy split into static info (basic settings), semi-static (e.g. mode) and dynamic info (currents)
- get rid of derived info, e.g. mode_id vs. mode and access bit
- how do we handle stuff like error-id vs error (string)? 
    a) we export both, so rfidreader=0 and rfidreaderstr="Not Installed"
    b) we only export the code, so rfidreader=0, and we make sure all API clients translate this to the right string; so index.html, ha-smartevse all have to know the right strings, how do we guard this
    c) we only export the string, so rfidreader = "Not Installed" , and all API clients will have to test for the strings. Spelling becomes important! mode = SMART, smart or Smart?

   Personally I currently lean towards c) , your thoughts?
- names in HA: keep the hierarchy in the (hidden) id's, so sensor.smartevse-5401.settings.mincurrent, but hide it in the name, so not "SmartEVSE Min Current" but Minimal Current (as probably "SmartEVSE" will be shown in the dasboard box already)  

Your thoughts?

**Comment by djoenez** [comment]
Hi,

I would also vote for C, but just set the naming once and never change it :) and your idea for HA is good, since I have 2 EVSEs I already have to change the naming anyway.

further improvements: documentation of MQTT and sending of currents via MQTT (Integration of MQTT into HA device?)

Another improvement: Buy me a coffee link as a show of appreciation, I would love that Ii could say that my C or python skills are good enough to help you guys, but this is not the case. So i would like to show my appreciation in monetary ways! :)

**Comment by stevoh6** [comment]
I will to with option c) ale all lowercase. But the documentation needs to be clear and include all possible strings. :)
